ZuluContent/Spells/Seventh/Polymorph.cs

```diff
@@ -73,164 +97,57 @@ public override bool CanCast()
-                {
-                    if (!((Body) body).IsHuman)
-                    {
-                        var mt = caster.Mount;
-
-                        if (mt != null)
-                            mt.Rider = null;
-                    }
-
-                    caster.BodyMod = body;
-
-                    if (body == 400 || body == 401)
-                        caster.HueMod = Race.DefaultRace.RandomSkinHue();
-                    else
-                        caster.HueMod = 0;
-
-                    BaseArmor.ValidateMobile(caster);
-                    BaseClothing.ValidateMobile(caster);
-
-                    StopTimer(caster);
-                    
-                    Timer t = new InternalTimer(caster, duration);
-                    Timers[caster] = t;
-
-                    t.Start();
-                    return true;
-                }
-            }
-            else
-            {
-                caster.SendLocalizedMessage(1005559); // This spell is already in effect.
-            }
-            return false;
-        }
+            if (!Caster.CanBuff(Caster, icons: BuffIcon.Polymorph))
+                return;
 
-        public static bool StopTimer(Mobile m)
-        {
-            var t = (Timer) Timers[m];
+            var magery = Caster.Skills[SkillName.Magery].Value;
 
-            if (t != null)
+            var group = Utility.RandomMinMax(0, 1);
+            var critter = Utility.RandomMinMax(0, 9);
+            
+            switch (magery)
             {
-                t.Stop();
-                Timers.Remove(m);
+                case < 80:
+                    group += 1;
+                    break;
+                case < 90:
+                    group += 2;
+                    break;
+                case >= 90:
+                    group = 3;
+                    break;
             }
 
-            return t != null;
-        }
-
-        private static void EndPolymorph(Mobile m)
-        {
-            if (!m.CanBeginAction(typeof(PolymorphSpell)))
+            if (Caster.GetClassBonus(SkillName.Magery) > 1.0)
             {
-                m.BodyMod = 0;
-                m.HueMod = -1;
-                m.EndAction(typeof(PolymorphSpell));
-
-                BaseArmor.ValidateMobile(m);
-                BaseClothing.ValidateMobile(m);
+                group += 1;
+                critter += 2;
             }
-        }
 
-        private class InternalTimer : Timer
-        {
-            private readonly Mobile m_Owner;
-
-            public InternalTimer(Mobile owner, TimeSpan duration) : base(duration)
-            {
-                m_Owner = owner;
-                Priority = TimerPriority.OneSecond;
-            }
+            group = Math.Clamp(group, 0, Groups.Length - 1);
+            critter = Math.Clamp(critter, 0, Groups[group].Length - 1);
+            
+            
+            // TODO: Use this for ebook Shape Shift?
+            // var bodyId = await new NewPolymorphGump(Caster);
```

Comment:
@michaelwhelehan left this here for you as an example for when you get to Shape Shift